sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,246 @@
+  }
+
+  if (!hmac->hmac_started) {
+    return kDifHmacFifoErrorInvalidState;
+  }
+
+  const uint8_t *data_sent = (const uint8_t *)data;
+
+  int unaligned_bytes = (uintptr_t)data_sent % sizeof(uint32_t);
+  int words_remaining = (len - unaligned_bytes) / sizeof(uint32_t);
+  int bytes_remaining = (len - unaligned_bytes) % sizeof(uint32_t);
+
+  // Bring |data_sent| into alignment so we can use |read32()| in the next loop.
+  while (unaligned_bytes > 0 && get_fifo_available_space(hmac) > 0) {
+    mmio_region_write8(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, *data_sent);
+    data_sent += sizeof(uint8_t);
+    --unaligned_bytes;
+  }
+
+  // Write whole words until the FIFO fills or there are no words left to send.
+  while (words_remaining > 0 && get_fifo_available_space(hmac) > 0) {
+    uint32_t word = read_32(data_sent);
+
+    mmio_region_write32(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, word);
+    data_sent += sizeof(uint32_t);
+    --words_remaining;
+  }
+
+  // Once all words are written write remaining bytes individually.
+  while (words_remaining == 0 && bytes_remaining > 0 &&
+         get_fifo_available_space(hmac) > 0) {
+    mmio_region_write8(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, *data_sent);
+    data_sent += sizeof(uint8_t);
+    --bytes_remaining;
+  }
+
+  if (bytes_sent != NULL) {
+    *bytes_sent = len - (words_remaining * sizeof(uint32_t)) - bytes_remaining;
+  }
+
+  if (words_remaining > 0 || bytes_remaining > 0) {
+    return kDifHmacFifoErrorFull;
+  }
+
+  return kDifHmacFifoErrorNone;
+}
+
+dif_hmac_error_t dif_hmac_fifo_count_entries(const dif_hmac_t *hmac,
+                                             uint32_t *num_entries) {
+  if (hmac == NULL || num_entries == NULL) {
+    return kDifHmacErrorInvalidArgument;
+  }
+
+  *num_entries = get_fifo_entry_count(hmac);
+
+  return kDifHmacErrorNone;
+}
+
+dif_hmac_error_t dif_hmac_get_message_length(const dif_hmac_t *hmac,
+                                             uint64_t *msg_len) {
+  if (hmac == NULL || msg_len == NULL) {
+    return kDifHmacErrorInvalidArgument;
+  }
+
+  uint64_t msg_lower =
+      mmio_region_read32(hmac->base_addr, HMAC_MSG_LENGTH_LOWER_REG_OFFSET);
+  uint64_t msg_upper =
+      mmio_region_read32(hmac->base_addr, HMAC_MSG_LENGTH_UPPER_REG_OFFSET);
+
+  *msg_len = (msg_upper << 32) | msg_lower;
+
+  return kDifHmacErrorNone;
+}
+
+dif_hmac_error_t dif_hmac_run(dif_hmac_t *hmac) {
+  if (hmac == NULL) {
+    return kDifHmacErrorInvalidArgument;
+  }
+
+  mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_CMD_REG_OFFSET,
+                                  HMAC_CMD_HASH_PROCESS);
+  hmac->hmac_started = false;
+
+  return kDifHmacErrorNone;
+}
+
+dif_hmac_error_t dif_hmac_done(const dif_hmac_t *hmac, bool *done) {
+  if (hmac == NULL || done == NULL) {
+    return kDifHmacErrorInvalidArgument;
+  }
+
+  // Check if hmac_done is asserted.
+  if (mmio_region_get_bit32(hmac->base_addr, HMAC_INTR_STATE_REG_OFFSET,
```

Comment:
This is basically code golf at this point but how do you feel about
```c
*done = get_bit32();
if(*done) {
  set_bit32();
}
```
Something about gift-wrapping a `= bool` feels gross to me.

Reply:
Something about declaring a new variable to only use it once feels gross to me. :wink:

Reply:
It's an out-param too, which is why its getting assigned in the first place.